File: backend/RAG/src/answer_generator_openrouter.py
import os
import time
import logging
from openai import OpenAI
from dotenv import load_dotenv

load_dotenv()

logger = logging.getLogger(__name__)


class AnswerGeneratorOpenRouter:
    """
    Answer Generator using OpenRouter API.
    
    This is an alternative to the Azure OpenAI-based answer generator,
    using OpenRouter's API to access various LLM models.
    """
    
    def __init__(self):
        self.api_key = os.getenv("OPENROUTER_API_KEY")
        self.model = os.getenv("OPENROUTER_MODEL", "openai/gpt-4o-mini")
        self.client = None
        self.system_prompt = """You are a helpful assistant that answers questions based on provided context.
Always use the provided context to answer questions accurately.
If the context doesn't contain enough information, say so clearly.
Keep answers concise and well-structured.
Format your answer in a clear, readable way."""

    def initialize(self):
        """Initialize OpenRouter client."""
        try:
            if not self.api_key:
                logger.error("OPENROUTER_API_KEY not found in environment variables")
                return False
            
            # Initialize OpenAI client with OpenRouter endpoint
            self.client = OpenAI(
                base_url="https://openrouter.ai/api/v1",
                api_key=self.api_key
            )
            
            logger.info(f"OpenRouter chat model '{self.model}' initialized.")
            return True
        except Exception as e:
            logger.error(f"Error initializing OpenRouter client: {e}")
            return False

    def set_system_prompt(self, prompt):
        """Set custom system prompt."""
        self.system_prompt = prompt

    def generate_answer(self, question, context_documents, max_tokens=1000):
        """Generate an answer based on retrieved documents using OpenRouter."""
        try:
            # Build context from documents
            context = self._build_context(context_documents)
            
            # Create messages
            messages = [
                {
                    "role": "system",
                    "content": self.system_prompt
                },
                {
                    "role": "user",
                    "content": f"""Based on the following context, answer the question.

CONTEXT:
{context}

QUESTION:
{question}

ANSWER:"""
                }
            ]
            
            # Generate response with streaming for accurate timing
            start_time = time.time()
            first_byte_time = None
            full_response = ""
            token_count = 0
            
            # Stream the response to capture first token timing
            response = self.client.chat.completions.create(
                model=self.model,
                messages=messages,
                max_tokens=max_tokens,
                stream=True,
                extra_headers={
                    "HTTP-Referer": "https://github.com/your-repo",  # Optional: for rankings
                    "X-Title": "Azure Voice Live RAG"  # Optional: for rankings
                }
            )
            
            for chunk in response:
                # Check if chunk has choices and content
                if chunk.choices and len(chunk.choices) > 0:
                    delta = chunk.choices[0].delta
                    if delta.content is not None:
                        # Record time of first token
                        if first_byte_time is None:
                            first_byte_time = time.time() - start_time
                        
                        # Accumulate the response
                        full_response += delta.content
                        token_count += 1
            
            total_time = time.time() - start_time
            
            # Log timing only if we got content
            if first_byte_time is not None:
                logger.info(f"[TIMING] OpenRouter LLM generation: First token in {first_byte_time:.2f}s, All tokens in {total_time:.2f}s ({token_count} tokens)")
            else:
                logger.warning(f"[TIMING] OpenRouter LLM generation: No tokens received in {total_time:.2f}s")
            
            # Return dict with answer and timing data
            return {
                "answer": full_response.strip(),
                "first_token_time": first_byte_time,
                "total_time": total_time,
                "token_count": token_count
            }
        except Exception as e:
            logger.error(f"Error generating answer with OpenRouter: {e}")
            return None

    def _build_context(self, documents):
        """Build context string from retrieved documents."""
        context_parts = []
        for i, doc in enumerate(documents, 1):
            context_parts.append(f"[Source {i}: {doc['source']}]\n{doc['content']}\n")
        return "\n".join(context_parts) if context_parts else "No relevant context found."

File: senacor-voicelive/lib/rag/answer-generator-openrouter.ts
/**
 * Answer Generator (OpenRouter)
 * Migrated from: backend/RAG/src/answer_generator_openrouter.py
 *
 * Alternative answer generator using OpenRouter API for various LLM models.
 */

import OpenAI from 'openai'
import type { SearchDocument } from './vector-search'
import type { AnswerResult } from './answer-generator'

export class AnswerGeneratorOpenRouter {
  private apiKey: string
  private model: string
  private client: OpenAI | null = null
  public systemPrompt: string

  constructor() {
    this.apiKey = process.env.OPENROUTER_API_KEY ?? ''
    this.model = process.env.OPENROUTER_MODEL ?? 'openai/gpt-4o-mini'
    this.systemPrompt = `You are a helpful assistant that answers questions based on provided context.
Always use the provided context to answer questions accurately.
If the context doesn't contain enough information, say so clearly.
Keep answers concise and well-structured.
Format your answer in a clear, readable way.`
  }

  /** Initialize OpenRouter client. */
  initialize(): boolean {
    try {
      if (!this.apiKey) {
        console.error('OPENROUTER_API_KEY not found in environment variables')
        return false
      }

      this.client = new OpenAI({
        baseURL: 'https://openrouter.ai/api/v1',
        apiKey: this.apiKey,
      })

      console.log(`OpenRouter chat model '${this.model}' initialized.`)
      return true
    } catch (e) {
      console.error(`Error initializing OpenRouter client: ${e}`)
      return false
    }
  }

  /** Set custom system prompt. */
  setSystemPrompt(prompt: string): void {
    this.systemPrompt = prompt
  }

  /** Generate an answer based on retrieved documents using OpenRouter (with streaming). */
  async generateAnswer(
    question: string,
    contextDocuments: SearchDocument[],
    maxTokens = 1000,
  ): Promise<AnswerResult | null> {
    try {
      if (!this.client) throw new Error('Not initialized')

      const context = this.buildContext(contextDocuments)

      const messages = [
        { role: 'system' as const, content: this.systemPrompt },
        {
          role: 'user' as const,
          content: `Based on the following context, answer the question.\n\nCONTEXT:\n${context}\n\nQUESTION:\n${question}\n\nANSWER:`,
        },
      ]

      const startTime = performance.now()
      let firstTokenTime: number | null = null
      let fullResponse = ''
      let tokenCount = 0

      const stream = await this.client.chat.completions.create({
        model: this.model,
        messages,
        max_completion_tokens: maxTokens,
        stream: true,
      })

      for await (const chunk of stream) {
        if (chunk.choices && chunk.choices.length > 0) {
          const delta = chunk.choices[0].delta
          if (delta?.content) {
            if (firstTokenTime === null) {
              firstTokenTime = performance.now() - startTime
            }
            fullResponse += delta.content
            tokenCount++
          }
        }
      }

      const totalTime = performance.now() - startTime

      if (firstTokenTime !== null) {
        console.log(
          `[TIMING] OpenRouter LLM generation: First token in ${(firstTokenTime / 1000).toFixed(2)}s, ` +
          `All tokens in ${(totalTime / 1000).toFixed(2)}s (${tokenCount} tokens)`,
        )
      } else {
        console.warn(`[TIMING] OpenRouter LLM generation: No tokens received in ${(totalTime / 1000).toFixed(2)}s`)
      }

      return {
        answer: fullResponse.trim(),
        first_token_time_ms: firstTokenTime ?? undefined,
        total_time_ms: totalTime,
        token_count: tokenCount,
      }
    } catch (e) {
      console.error(`Error generating answer with OpenRouter: ${e}`)
      return null
    }
  }

  /** Build context string from retrieved documents. */
  private buildContext(documents: SearchDocument[]): string {
    const parts = documents.map(
      (doc, i) => `[Source ${i + 1}: ${doc.source}]\n${doc.content}\n`,
    )
    return parts.length > 0 ? parts.join('\n') : 'No relevant context found.'
  }
}
